Kp/ll stacks queues (#24)
Ada-Developers-Academy/ada-CS-Fundamentals-Internship: 02-linked-lists/04-linked-lists-supplemental-concepts.md

# Supplemental Concepts

## The Call Stack

As methods get called in an application, the system stores the current instruction addresses, local variables etc on a stack known as the _system call stack_.  Then when a method ends, the topmost method is popped off the stack allowing the system to resume execution.

![Memory Layout Diagram](images/memory-layout.png)

The diagram below shows the memory used by a running application.  At the top, the text in the application, and global variables are stored.  Below them is the dyanmic data allocated.  At the bottom the system stack stores the functions called.

```python
def function_a(x):
    y = 4
    z = function_b(x, y)
    print(f"The number is {z}")

def function_b(x, y):
  # pause application
    return x + y

x = 3
function_a(x)
```

For the code snippet above, the stack frame at the `# pause application` line would look like this:

![Stack frame example](images/call-stack.png)

<!-- Graphic saved at: https://drive.google.com/file/d/1SsxrrqIIw5oYonpKzwxvXAF3lTcQmB10/view?usp=sharing -->

> As the application starts the system puts the main part of the application on the stack, and then main calls `function_a`.  So the system saves the arguments to `function_a` onto the stack, the address to return to (Main) when the method is finished and the local variabes in `function_a`.  Then `function_b` is called and the system saves the arguments to `function_b` onto the stack, the return address (function_a) and any local variables in `function_b`.  When function_b finishes, it's stack frame is popped off the stack, and the return address is used to resume `function_a`.  When `function_a` finishes the same pop operation is performed and the application returns to main.  When the main part of the application is finished, it's stack frame is popped off and the application terminates.  

Reviewer: Oh man, this is a doozy. It was a little difficult for me to follow along. I might recommend we break this up with bullet points for digestibility. 

Also, I think the graphic may have a small typo in the stack frame for function_b: should it say "Return address function_a" instead of "Return address function_b"?

> As the application starts the system puts the main part of the application on the stack, and then main calls `function_a`.  So the system saves the arguments to `function_a` onto the stack, the address to return to (Main) when the method is finished and the local variabes in `function_a`.  Then `function_b` is called and the system saves the arguments to `function_b` onto the stack, the return address (function_a) and any local variables in `function_b`.  When function_b finishes, it's stack frame is popped off the stack, and the return address is used to resume `function_a`.  When `function_a` finishes the same pop operation is performed and the application returns to main.  When the main part of the application is finished, it's stack frame is popped off and the application terminates.  

Reviewer: ```suggestion
- As the application starts the system puts the main part of the application on the stack, and then main calls `function_a`.  
- The system saves the arguments to `function_a` onto the stack, the address to return to (Main) when the method is finished and the local variabes in `function_a`.  
- `function_b` is called and the system saves the arguments to `function_b` onto the stack, the return address (`function_a`) and any local variables in `function_b`.  
- When `function_b` finishes, its stack frame is popped off the stack, and the return address is used to resume `function_a`.  
- When `function_a` finishes the same pop operation is performed and the application returns to main.  
- When the main part of the application is finished, its stack frame is popped off and the application terminates.  

```
